### ISSUE TITLE
Remove or downgrade "Use a non-snapshot version number for dependency"

### ISSUE THREAD

**Issue by joewreschnig** [purcell/package-lint#99]
I think this advice is not helpful, and is probably actively harmful.

I don't have much control over how my dependencies are handling releases. If they only have snapshots, or use snapshot versions, I must depend on a snapshot. So this is not actionable advice for the author of the package being checked. If we really want to discourage versioning like this, the warning should be on declaring a snapshot version, not using one.

But I think date-based versioning is probably the best option package authors have right now, as long as `version-list-<=` is the only dependency relation Emacs understands. If I am using a yet-unreleased feature I need to depend on a snapshot version, even when my dependencies are tagging releases. And with a date-based dependency the depender doesn't need to be updated again once its dependencies "go gold".

The risk of such things is that an unstable feature might "go away" but Emacs's naive version dependency resolution means I can't avoid this risk with any versioning scheme.

**Comment by purcell** [comment]
Yeah, whether or not to use snapshot dependency is certainly a nuanced issue, and the above is an accurate summary. What might be nice would be to make it a warning (if it's currently an error), and I could imagine that trade-offs like you outline could be documented on, say, a wiki page here so that authors could read more about the rationale.

**Comment by joewreschnig** [comment]
> What might be nice would be to make it a warning (if it's currently an error)

It's currently a warning, but warnings cause `package-lint-batch-and-exit` to exit non-zero, so it's hard to make a functional CI pipeline for something with a snapshot dependency.

**Comment by dickmao** [comment]
I feel package-requires should *only* specify snapshot versions for melpa dependencies (so quite the opposite of what this safeguard intends).  My experience is that specifying a latest-greatest non-snapshot dependency such as `(dash "99.99.alpha")` doesn't cause `package.el` to update dash because an antiquated melpa-imposed version like "19691231" is still considered "greater."

**Comment by purcell** [comment]
> I feel package-requires should _only_ specify snapshot versions for melpa dependencies (so quite the opposite of what this safeguard intends). My experience is that specifying a latest-greatest non-snapshot dependency such as `(dash "99.99.alpha")` doesn't cause `package.el` to update dash because an antiquated melpa-imposed version like "19691231" is still considered "greater."

@dickmao That just makes your source package uninstallable by MELPA Stable users who don't also use MELPA. The general solution for the version snapshot problem is https://github.com/melpa/melpa/issues/2955, btw.

**Comment by tsuu32** [comment]
> It's currently a warning, but warnings cause package-lint-batch-and-exit to exit non-zero, so it's hard to make a functional CI pipeline for something with a snapshot dependency.

Hi,
why does package-lint exit Emacs with non-zero even when warning?

**Comment by purcell** [comment]
> > 
> 
> Hi,
> why does package-lint exit Emacs with non-zero even when warning?

This can now be controlled, which somewhat mitigates the overall issue.

**Comment by tsuu32** [comment]
@purcell
Oh, I didn't notice Issue #163.
Thanks!
